File: web/src/api/v2/watchTogetherMemberState.ts
import type { components } from "./schema";
import {
  captureProfileRequestContext,
  isCapturedProfileAuthorityActive,
  StaleApiRequestContextError,
} from "@/api/client";
import { v2 } from "./request";

export type MemberWatchStateKind = "unseen" | "in_progress" | "watched";

export interface MemberWatchState {
  user_id: number;
  profile_id: string;
  state: MemberWatchStateKind;
  position_seconds?: number;
  duration_seconds?: number;
  on_watchlist: boolean;
}

export interface ItemMemberState {
  content_id: string;
  members: MemberWatchState[];
}

export interface RoomMemberStateResponse {
  members: components["schemas"]["WatchTogetherRoomMember"][];
  items: ItemMemberState[];
}

/** The server bound on one member-state read. */
export const MAX_MEMBER_STATE_IDS = 200;

function numericID(value: string): number {
  const id = Number(value);
  if (!/^[1-9][0-9]*$/.test(value) || !Number.isSafeInteger(id))
    throw new Error("Invalid member identity.");
  return id;
}

/**
 * Classifies the named content per connected room member. A POST-shaped read;
 * the server changes nothing.
 */
export async function queryRoomMemberState(
  roomId: string,
  roomToken: string,
  contentIds: string[],
  authority = captureProfileRequestContext(),
): Promise<RoomMemberStateResponse> {
  if (!authority || !isCapturedProfileAuthorityActive(authority))
    throw new StaleApiRequestContextError();
  const ids = Array.from(new Set(contentIds.map((id) => id.trim()).filter(Boolean)));
  if (ids.length === 0) return { members: [], items: [] };
  if (ids.length > MAX_MEMBER_STATE_IDS)
    throw new Error(`Member state reads are limited to ${MAX_MEMBER_STATE_IDS} items.`);
  const result = await v2("POST /api/v2/watch-together/rooms/{room_id}/member-state", {
    path: { room_id: roomId },
    headers: { "X-Room-Token": roomToken },
    body: { content_ids: ids },
    profileContext: authority,
    retryAuthentication: false,
  }).catch((error: unknown) => {
    if (!isCapturedProfileAuthorityActive(authority)) throw new StaleApiRequestContextError();
    throw error;
  });
  if (!isCapturedProfileAuthorityActive(authority)) throw new StaleApiRequestContextError();
  return {
    members: result.members,
    items: result.items.map((item) => ({
      content_id: item.content_id,
      members: item.members.map((member) => ({
        user_id: numericID(member.user_id),
        profile_id: member.profile_id,
        state: member.state,
        position_seconds: member.position_seconds,
        duration_seconds: member.duration_seconds,
        on_watchlist: member.on_watchlist,
      })),
    })),
  };
}

File: internal/apiv2/watch_together_member_state.go
package apiv2

import (
	"context"
	"net/http"
	"strconv"

	"github.com/Silo-Server/silo-server/internal/api/handlers"
	catalogpkg "github.com/Silo-Server/silo-server/internal/catalog"
	"github.com/Silo-Server/silo-server/internal/watchtogether"
)

// WatchTogetherMemberStateService answers per-member watch state for content
// the caller may access. The member set comes from the shared room runtime.
type WatchTogetherMemberStateService interface {
	CheckSuggestionRoomProof(string, int, string, string) error
	RoomMemberState(context.Context, string, int, string, []string, catalogpkg.AccessFilter) ([]watchtogether.MemberSummary, []watchtogether.ItemMemberState, error)
}
type WatchTogetherMemberStateInput struct {
	RoomID    string `path:"room_id"`
	RoomToken string `header:"X-Room-Token" maxLength:"4096"`
	Body      struct {
		ContentIDs []ID `json:"content_ids" minItems:"1" maxItems:"200"`
	}
}
type WatchTogetherMemberWatchState struct {
	UserID          ID       `json:"user_id"`
	ProfileID       string   `json:"profile_id"`
	State           string   `json:"state" enum:"unseen,in_progress,watched"`
	PositionSeconds *float64 `json:"position_seconds,omitempty"`
	DurationSeconds *float64 `json:"duration_seconds,omitempty"`
	OnWatchlist     bool     `json:"on_watchlist"`
}
type WatchTogetherItemMemberState struct {
	ContentID ID                              `json:"content_id"`
	Members   []WatchTogetherMemberWatchState `json:"members"`
}
type WatchTogetherMemberState struct {
	Members []WatchTogetherRoomMember      `json:"members" doc:"Connected room members across API servers, host first"`
	Items   []WatchTogetherItemMemberState `json:"items" doc:"One entry per distinct accessible requested content id, in request order"`
}
type WatchTogetherMemberStateOutput struct {
	CacheControl string `header:"Cache-Control"`
	Body         WatchTogetherMemberState
}

func registerWatchTogetherMemberState(reg *Registry) {
	op := Operation{Operation: humaOp(http.MethodPost, Prefix+"/watch-together/rooms/{room_id}/member-state", "queryWatchTogetherMemberState", "realtime", "Read what each connected room member has watched of the named content: unseen, in progress (with position) or watched, and whether it is on their watchlist. A POST-shaped read: the id set exceeds what a query string carries. Members across API servers are read; inaccessible content ids are omitted."), Class: ClassProfileScoped, ServiceBacked: true, RetrySafety: RetrySafetyNaturalIdempotent}
	op.MaxBodyBytes = 16384
	op.Errors = []int{409}
	Register(reg, op, func(ctx context.Context, in *WatchTogetherMemberStateInput) (*WatchTogetherMemberStateOutput, error) {
		svc := reg.deps.WatchTogetherMemberState
		if svc == nil {
			return nil, unavailable("watch together")
		}
		user, profile, p := viewerIdentity(ctx)
		if p != nil {
			return nil, p
		}
		if err := svc.CheckSuggestionRoomProof(in.RoomID, user, profile, in.RoomToken); err != nil {
			return nil, serviceProblem(err)
		}
		if reg.deps.CatalogAccess == nil {
			return nil, unavailable("catalog access")
		}
		filter, err := reg.deps.CatalogAccess.ContextAccessFilter(ctx, handlers.AccessFilterOptions{})
		if err != nil {
			return nil, collectionProblem(err)
		}
		ids := make([]string, 0, len(in.Body.ContentIDs))
		for i, id := range in.Body.ContentIDs {
			if id == "" {
				return nil, NewProblem(TypeValidationFailed, "Content ids must be non-empty.").WithErrors(ProblemError{Location: "body.content_ids[" + strconv.Itoa(i) + "]", Code: codeInvalid})
			}
			ids = append(ids, string(id))
		}
		members, items, err := svc.RoomMemberState(ctx, in.RoomID, user, profile, ids, filter)
		if err != nil {
			return nil, suggestionProblem(err)
		}
		out := &WatchTogetherMemberStateOutput{CacheControl: adminLogsSocketCacheControl}
		out.Body.Members = watchTogetherMembersOf(members)
		out.Body.Items = make([]WatchTogetherItemMemberState, 0, len(items))
		for _, item := range items {
			entry := WatchTogetherItemMemberState{ContentID: ID(item.ContentID), Members: make([]WatchTogetherMemberWatchState, 0, len(item.Members))}
			for _, m := range item.Members {
				entry.Members = append(entry.Members, WatchTogetherMemberWatchState{UserID: IDFromInt(int64(m.UserID)), ProfileID: m.ProfileID, State: string(m.State), PositionSeconds: m.PositionSeconds, DurationSeconds: m.DurationSeconds, OnWatchlist: m.OnWatchlist})
			}
			out.Body.Items = append(out.Body.Items, entry)
		}
		return out, nil
	})
}

func watchTogetherMembersOf(rows []watchtogether.MemberSummary) []WatchTogetherRoomMember {
	out := make([]WatchTogetherRoomMember, 0, len(rows))
	for _, m := range rows {
		out = append(out, WatchTogetherRoomMember{UserID: IDFromInt(int64(m.UserID)), ProfileID: m.ProfileID, DisplayName: m.DisplayName, IsHost: m.IsHost, IsSelf: m.IsSelf, Connected: m.Connected, LobbyReady: m.LobbyReady})
	}
	return out
}
